### PR TITLE
[Flink] fix data overflow while convert Integer to Long (#3738)

### PR DESCRIPTION
**Pull request by chenzihao5**
1. What this MR does ?
    Fix data overflow while writing data whose type is java.sql.Time.
2. CHANGELOG
    Flink version: 1.12, 1.13, 1.14
    function: org.apache.iceberg.flink.data.FlinkValueWriters.TimeMicrosWriter#write

@openinx @JingsongLi Please help to confirm this Problem. Thanks very much.

### CODE REVIEW

**Review by kbendick — COMMENTED** [review]
I checked (with a Scala shell but using `java.lang.Integer`) and using the float does indeed prevent overflow.

Could you add test cases somewhere for writing a value that would have overflowed otherwise?

**Comment by chenzihao5** [comment]
> Thanks for this patch @chenzihao5!
> 
> I checked (with a Scala shell but using `java.lang.Integer`) and using the explicit float does indeed seem to prevent overflow.
> 
> ```
> scala> import java.lang.{Integer => JInt}
> import java.lang.{Integer=>JInt}
> 
> scala> JInt.MAX_VALUE
> res5: Int = 2147483647
> 
> scala> JInt.MAX_VALUE * 1000
> res6: Int = -1000
> 
> scala> JInt.MAX_VALUE * 1000L
> res7: Long = 2147483647000
> ```
> 
> Could you add test cases somewhere for writing a value that would have overflowed otherwise?

@kbendick Thanks for your review. I will add some unit tests.

**Comment by openinx** [comment]
Great finding, @chenzihao5 . It's indeed an overflow bug !

**Comment by rdblue** [comment]
I'm marking this for inclusion in 0.13.0. We should wait for tests, but I want to make sure we include this in the release.

**Comment by chenzihao5** [comment]
> I'm marking this for inclusion in 0.13.0. We should wait for tests, but I want to make sure we include this in the release.

Thanks for reminding, I will complete the unit test today.

**Comment by chenzihao5** [comment]
@rdblue @kbendick @openinx Hi, All. I have added unit tests in version 1.14, please help to review it, Thanks.

About these unit tests:
    1. test for numeric type, such as: int, float, double, date, time, timestamp, long, decimal.
    2. test MAX_VALUE and MIN_VALUE of all types.
    3. If this version is completed, I will add unit tests in version 1.12 and 1.13 too.

Thanks for reviewing!

File: flink/v1.14/flink/src/test/java/org/apache/iceberg/flink/sink/TestFlinkIcebergSinkNumData.java [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+import java.sql.Date;
+import java.sql.Time;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+@RunWith(Parameterized.class)
+public class TestFlinkIcebergSinkNumData extends TableTestBase {
+    @ClassRule
```
Comment (rdblue) [inline flink/v1.14/flink/src/test/java/org/apache/iceberg/flink/sink/TestFlinkIcebergSinkNumData.java:65]: Iceberg uses 2 spaces for an indent. Can you please reformat this file?

File: flink/v1.14/flink/src/test/java/org/apache/iceberg/flink/sink/TestFlinkIcebergSinkNumData.java [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+                new Object[]{"avro", 2, true},
+                new Object[]{"avro", 2, false},
+                new Object[]{"orc", 1, true},
+                new Object[]{"orc", 1, false},
+                new Object[]{"orc", 2, true},
+                new Object[]{"orc", 2, false},
+                new Object[]{"parquet", 1, true},
+                new Object[]{"parquet", 1, false},
+                new Object[]{"parquet", 2, true},
+                new Object[]{"parquet", 2, false}
```
[the remark this answers is not in the document]
Comment (chenzihao5) [reply flink/v1.14/flink/src/test/java/org/apache/iceberg/flink/sink/TestFlinkIcebergSinkNumData.java:112]: Thanks. I will update it.

**Comment by chenzihao5** [comment]
@rdblue I have updated test into existing test suite for Avro writer and reformat code style. Please help to review it.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @chenzihao5! Looks good to me now. I'm running CI and will merge when it passes.

**Comment by rdblue** [comment]
Thanks, @chenzihao5! I merged this.

**Pull request merged**